wandb/sdk/internal/system/assets/gpu_amd.py

```diff
@@ -0,0 +1,190 @@
+
+            for key in samples[0].keys():
+                samples_key = [s[key] for s in samples]  # type: ignore
+                aggregate = aggregate_mean(samples_key)
+                stats[self.name.format(gpu_id=i, key=key)] = aggregate
+
+        return stats
+
+
+@asset_registry.register
+class GPUAMD:
+    """GPUAMD is a class for monitoring AMD GPU devices.
+
+    Uses AMD's rocm_smi tool to get GPU stats.
+    For the list of supported environments and devices, see
+    https://github.com/RadeonOpenCompute/ROCm/blob/develop/docs/deploy/
+    """
+
+    def __init__(
+        self,
+        interface: "Interface",
+        settings: "SettingsStatic",
+        shutdown_event: threading.Event,
+    ) -> None:
+        self.name = self.__class__.__name__.lower()
+        self.metrics: List[Metric] = [
+            GPUAMDStats(),
+        ]
+        self.metrics_monitor = MetricsMonitor(
+            self.name,
+            self.metrics,
+            interface,
+            settings,
+            shutdown_event,
+        )
+        telemetry_record = telemetry.TelemetryRecord()
+        telemetry_record.env.amd_gpu = True
+        interface._publish_telemetry(telemetry_record)
+
+    @classmethod
+    def is_available(cls) -> bool:
+        try:
+            rocm_smi_available = shutil.which(ROCM_SMI_CMD) is not None
+            if rocm_smi_available:
+                _ = get_rocm_smi_stats()
+                return True
+        except Exception:
+            pass
+        return False
+
+    def start(self) -> None:
+        self.metrics_monitor.start()
+
+    def finish(self) -> None:
+        self.metrics_monitor.finish()
+
+    def probe(self) -> dict:
+        info: _InfoDict = {}
+        try:
+            stats = get_rocm_smi_stats()
+
+            info["gpu_count"] = len(
+                [key for key in stats.keys() if key.startswith("card")]
+            )
+            info["gpu_devices"] = [
+                {
+                    "id": stats[key]["GPU ID"],
+                    "unique_id": stats[key]["Unique ID"],
+                    "vbios_version": stats[key]["VBIOS version"],
+                    "performance_level": stats[key]["Performance Level"],
+                    "gpu_overdrive": stats[key]["GPU OverDrive value (%)"],
+                    "gpu_memory_overdrive": stats[key][
+                        "GPU Memory OverDrive value (%)"
+                    ],
+                    "max_power": stats[key]["Max Graphics Package Power (W)"],
+                    "series": stats[key]["Card series"],
+                    "model": stats[key]["Card model"],
+                    "vendor": stats[key]["Card vendor"],
+                    "sku": stats[key]["Card SKU"],
+                    "sclk_range": stats[key]["Valid sclk range"],
+                    "mclk_range": stats[key]["Valid mclk range"],
+                }
+                for key in stats.keys()
+                if key.startswith("card")
+            ]
```

Comment:
nit: more pythonic to move this into an `else` block.  Reserve `try` for just the part that can raise exception

Reply:
all of this can raise an exception, that's why I did it this way -- there are gpus that rocm-smi doesn't fully support and so __getitem__ calls might fail. but let me rewrite it in a bit more conservative way.